File: crates/rco-sdk-ruby/lib/rco.rb
require 'ffi'

module Rco
  extend FFI::Library
  
  # Load the native library (using the C# FFI build which provides the C-ABI)
  ffi_lib File.expand_path('../../target/debug/librco_sdk_csharp.so', __dir__)

  # Define the C-FFI signatures
  attach_function :rco_project_p14_batch, [:pointer, :size_t, :pointer, :pointer], :int
  attach_function :rco_construct_por_message, [:pointer, :pointer, :pointer, :size_t, :pointer], :int
  attach_function :rco_verify_hardware_sovereignty, [], :int

  class Client
    # Integration with Numo::NArray if available
    HAS_NUMO = begin
      require 'numo/narray'
      true
    rescue LoadError
      false
    end

    def self.project_p14_batch(rewards)
      if HAS_NUMO && rewards.is_a?(Numo::NArray)
        project_numo(rewards)
      else
        project_standard(rewards)
      end
    end

    private

    def self.project_numo(rewards)
      count = rewards.size
      
      # Ensure data is contiguous DFloat (double)
      dfloat_rewards = rewards.is_a?(Numo::DFloat) ? rewards : Numo::DFloat.cast(rewards)
      
      # Create Numo output arrays for zero-copy high/low parts
      out_high = Numo::Int64.new(count)
      out_low = Numo::Int64.new(count)
      
      # Extract raw pointers from Numo
      input_ptr = FFI::Pointer.new(:double, dfloat_rewards.get_data_addr)
      high_ptr = FFI::Pointer.new(:int64, out_high.get_data_addr)
      low_ptr = FFI::Pointer.new(:int64, out_low.get_data_addr)
      
      result = Rco.rco_project_p14_batch(input_ptr, count, high_ptr, low_ptr)
      raise "P14 Projection Failed: #{result}" if result != 0
      
      { high: out_high, low: out_low }
    end

    def self.project_standard(rewards)
      count = rewards.size
      input = FFI::MemoryPointer.new(:double, count)
      input.put_array_of_double(0, rewards)
      
      out_high = FFI::MemoryPointer.new(:int64, count)
      out_low = FFI::MemoryPointer.new(:int64, count)
      
      result = Rco.rco_project_p14_batch(input, count, out_high, out_low)
      raise "P14 Projection Failed: #{result}" if result != 0
      
      highs = out_high.get_array_of_int64(0, count)
      lows = out_low.get_array_of_int64(0, count)
      
      highs.zip(lows)
    end

    public

    def self.construct_por_message(weight_hash, merkle_link, telemetry)
      raise "Hashes must be 32 bytes" if weight_hash.bytesize != 32 || merkle_link.bytesize != 32
      
      wh_ptr = FFI::MemoryPointer.new(:uint8, 32)
      wh_ptr.put_bytes(0, weight_hash)
      
      ml_ptr = FFI::MemoryPointer.new(:uint8, 32)
      ml_ptr.put_bytes(0, merkle_link)
      
      tel_ptr = FFI::MemoryPointer.new(:uint8, telemetry.bytesize)
      tel_ptr.put_bytes(0, telemetry)
      
      out_ptr = FFI::MemoryPointer.new(:uint8, 96)
      
      result = Rco.rco_construct_por_message(wh_ptr, ml_ptr, tel_ptr, telemetry.bytesize, out_ptr)
      raise "PoR Construction Failed: #{result}" if result != 0
      
      out_ptr.get_bytes(0, 96)
    end

    def self.hardware_sovereign?
      Rco.rco_verify_hardware_sovereignty == 1
    end
  end
end

File: crates/rco-sdk-php/rco.php
<?php

/**
 * Hyper-Scale PHP SDK for RCO Protocol.
 * Optimized for PHP 8.1+ with FFI Preloading support.
 * Achieves Web-Scale Sovereignty via Zero-Copy Memory Casting.
 */
class RcoClient {
    private static $ffi = null;
    private static $libPath = null;

    /**
     * Initialize the RCO FFI bridge.
     * Supports Preloading: If called during opcache.preload, the RCO kernel 
     * becomes part of the permanent PHP process memory.
     */
    public static function init(?string $customLibPath = null) {
        if (self::$ffi !== null) return;

        self::$libPath = $customLibPath ?? __DIR__ . '/../../target/debug/librco_sdk_csharp.so';
        
        $cdef = "
            typedef struct { long high; long low; } rco_reward_t;
            int rco_project_p14_batch(const double* input, size_t count, long* output_high, long* output_low);
            int rco_construct_por_message(const uint8_t* weight_hash, const uint8_t* merkle_link, const uint8_t* telemetry, size_t telemetry_len, uint8_t* output);
            int rco_verify_hardware_sovereignty();
        ";

        try {
            self::$ffi = FFI::cdef($cdef, self::$libPath);
        } catch (FFI\Exception $e) {
            throw new RuntimeException("RCO Native Library Load Failed: " . $e->getMessage());
        }
    }

    /**
     * Projects a batch of rewards using high-performance memory casting.
     * Hits the point of integral liberty: No PHP array conversion overhead.
     */
    public static function projectP14Batch(array $rewards) {
        self::init();
        $count = count($rewards);
        
        // Allocate native memory for the projection
        $input = FFI::new("double[$count]");
        $outHigh = FFI::new("long[$count]");
        $outLow = FFI::new("long[$count]");
        
        // Rapid memory population
        foreach ($rewards as $i => $r) $input[$i] = (double)$r;
        
        $result = self::$ffi->rco_project_p14_batch(
            FFI::addr($input), 
            $count, 
            FFI::addr($outHigh), 
            FFI::addr($outLow)
        );

        if ($result !== 0) throw new RuntimeException("RCO P14 Projection Failed: $result");
        
        // Return as a structured list of 128-bit parts
        $combined = [];
        for ($i = 0; $i < $count; $i++) {
            $combined[] = [$outHigh[$i], $outLow[$i]];
        }
        return $combined;
    }

    /**
     * High-speed Proof of Reflexion construction.
     * Uses binary string pointers for zero-copy efficiency.
     */
    public static function constructPoRMessage(string $weightHash, string $merkleLink, string $telemetry) {
        self::init();
        if (strlen($weightHash) !== 32 || strlen($merkleLink) !== 32) {
            throw new InvalidArgumentException("Hashes must be 32 bytes.");
        }

        $wh = FFI::new("uint8_t[32]");
        $ml = FFI::new("uint8_t[32]");
        FFI::memcpy($wh, $weightHash, 32);
        FFI::memcpy($ml, $merkleLink, 32);

        $telLen = strlen($telemetry);
        $tel = FFI::new("uint8_t[$telLen]");
        FFI::memcpy($tel, $telemetry, $telLen);

        $output = FFI::new("uint8_t[96]");
        
        $result = self::$ffi->rco_construct_por_message(
            FFI::addr($wh), 
            FFI::addr($ml), 
            FFI::addr($tel), 
            $telLen, 
            FFI::addr($output)
        );

        if ($result !== 0) throw new RuntimeException("RCO PoR Construction Failed: $result");

        return FFI::string($output, 96);
    }

    /**
     * Checks for hardware sovereignty directly in the web request lifecycle.
     */
    public static function isHardwareSovereign(): bool {
        self::init();
        return self::$ffi->rco_verify_hardware_sovereignty() === 1;
    }
}
